internal/gke/update.go

```diff
@@ -373,22 +375,37 @@ func UpdateLabels(
 	if config.Spec.Labels == nil || reflect.DeepEqual(config.Spec.Labels, upstreamSpec.Labels) || (upstreamSpec.Labels == nil && len(config.Spec.Labels) == 0) {
```

Comment:
> This change adds a retry to compensate for the fact that UpdateLabels is being called twice.

Can you explain that part, is it because the reconciling loop hits it again (because checkAndUpdate got entered) ? Or are you calling it twice from the code.

Reply:
Yeah so what's happening is we go into `checkAndUpdate` for the first time [here](https://github.com/rancher/gke-operator/blob/d57ee5fb1899cf8cdf7cab1be694c9d84c3e50d7/controller/gke-cluster-config-handler.go#L106) where gkeClusterConfig is in "active" phase. From there it goes to [updateUpstreamClusterState](https://github.com/rancher/gke-operator/blob/d57ee5fb1899cf8cdf7cab1be694c9d84c3e50d7/controller/gke-cluster-config-handler.go#L275) and then to [UpdateLabels](https://github.com/rancher/gke-operator/blob/d57ee5fb1899cf8cdf7cab1be694c9d84c3e50d7/controller/gke-cluster-config-handler.go#L364). Without this change, it successfully updates the cluster labels through the GKE API. Then it calls `enqueueUpdate` and sets status to [updating](https://github.com/rancher/gke-operator/blob/d57ee5fb1899cf8cdf7cab1be694c9d84c3e50d7/controller/gke-cluster-config-handler.go#L287). On the next reconciler loop, status is "updating" and so we go into [checkAndUpdate](https://github.com/rancher/gke-operator/blob/d57ee5fb1899cf8cdf7cab1be694c9d84c3e50d7/controller/gke-cluster-config-handler.go#L108) again. In any other update, we would hit [this](https://github.com/rancher/gke-operator/blob/d57ee5fb1899cf8cdf7cab1be694c9d84c3e50d7/controller/gke-cluster-config-handler.go#L241) and be bumped out of that function until the next loop, but GKE either never sets the cluster to reconciling or does it too quickly to be useful, and so we go back to `UpdateLabels`. But the cluster is actually still in the middle of setting the labels, and so the upstreamSpec labels check [here](https://github.com/rancher/gke-operator/blob/d57ee5fb1899cf8cdf7cab1be694c9d84c3e50d7/internal/gke/update.go#L373) is wrong, and it tries to go ahead with [the update](https://github.com/rancher/gke-operator/blob/d57ee5fb1899cf8cdf7cab1be694c9d84c3e50d7/internal/gke/update.go#L381-L391) and that's where the GKE API rejects it.

Reply:
We encountered something simillar in EKS and the way that we handled it is we allowed the controller to resend the update and when we evaluated the error as indicating an update was underway, we simple did not write the error to status: https://github.com/rancher/eks-operator/blob/master/controller/eks-cluster-config-handler.go#L111. This way the cluster will still be in reenqueued and that error will be as good as observing UPDATING/RECONCILING on the cluster. I also wonder if this could be caused by threadiness being higher than it needs to be here: https://github.com/rancher/gke-operator/blob/master/main.go#L67. I'm not sure but this may cause the framework to reenqueue the object while the currently running instance is not finished which would cause conflicts and race conditions.

Reply:
>  I also wonder if this could be caused by threadiness being higher than it needs to be

I suspected that too but changing it didn't seem to fix things